### PR TITLE
passwd plugin

### PR DESCRIPTION
**Pull request by tom-wa**
will add unittests later

can someone with a system not supporting fgetpwent/putpwent test the cmake code ?

EDITED

### CODE REVIEW

**Comment by markus2330** [comment]
Testing it would be much easier with unit tests ;)

When do you plan to add them? Within this PR?

Can you maybe add some simple shell recorder tests?

**Comment by markus2330** [comment]
Btw. about the system not supporting  fgetpwent/putpwent: do you want that the cmake code to be tested or the plugin itself? (I think you mean the cmake code, but your question suggests otherwise.)

**Comment by tom-wa** [comment]
you are right, just the cmake code needs to be tested( i couldn't find any consistent of checking if the functions exist other than try_compile), the rest should work as expected. 

if that pr gets merged without unittests or some shellrecorder test i'll probably never do them, so .. i'll add them to this PR later

File: src/plugins/passwd/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+include (LibAddMacros)
+try_compile(HAS_FGETPWENT "${CMAKE_BINARY_DIR}" "${PROJECT_SOURCE_DIR}/src/plugins/passwd/hasfgetpwent.c")
+if(HAS_FGETPWENT)
+    add_definitions(-DHAS_FGETPWENT)
+endif()
+try_compile(HAS_PUTPWENT "${CMAKE_BINARY_DIR}" "${PROJECT_SOURCE_DIR}/src/plugins/passwd/hasputpwent.c")
+if(HAS_PUTPWENT)
+    add_definitions(-DHAS_PUTPWENT)
```
Comment (markus2330) [inline src/plugins/passwd/CMakeLists.txt]: Please do not add them globally, use `COMPILE_DEFINITIONS` as argument in `add_plugin`.

(In one phase would be enough, I think `try_compile` in this setup will be executed multiple times?)

File: src/plugins/passwd/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+- infos/status = recommended productive maintained reviewed conformant compatible coverage specific unittest shelltest tested nodep libc configurable final preview memleak experimental difficult unfinished old nodoc concept orphan obsolete discouraged -1000000
+- infos/metadata =
+- infos/description =
+
+## Usage ##
+
+```
+kdb mount /etc/passwd system/passwd passwd index=name
+```
+if `index` is set to `name` passwd entrys will be sorted by name, if not set or set to `uid` passwd entry will be sorted by uid
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: Please document the fields your plugin will generate

File: src/plugins/passwd/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+- infos/recommends =
+- infos/placements = getstorage setstorage
+- infos/status = recommended productive maintained reviewed conformant compatible coverage specific unittest shelltest tested nodep libc configurable final preview memleak experimental difficult unfinished old nodoc concept orphan obsolete discouraged -1000000
+- infos/metadata =
+- infos/description =
+
+## Usage ##
+
+```
+kdb mount /etc/passwd system/passwd passwd index=name
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: Please copy shell recorder tests here (hopefully the shell recorder can soon execute parts of README.md)

**Comment by markus2330** [comment]
Yes, I think the same goes for docu, please add a bit more about how to use the plugin.

**Comment by markus2330** [comment]
I like the idea of `index`, it sounds very useful. Maybe we should add support for mounting the same file multiple times to make this idea even more powerful. But maybe transformations in separate plugins are better after all: it would keep the resolver simpler.

Are you sure that there is a system that has `pwd.h` and struct passwd defined but does not have putpwent and fgetpwent?

**Comment by tom-wa** [comment]
>Are you sure that there is a system that has pwd.h and struct passwd defined but does not have putpwent and fgetpwent?

openbsd. 
there are some others too but i'm not really sure either. it just says not posix everywhere

**Comment by tom-wa** [comment]
> But maybe transformations in separate plugins are better after all: it would keep the resolver simpler.

you are right. i'd say plugins use array notation für every line, and the transformation plugin .. does the transformation.

**Comment by markus2330** [comment]
The augeas plugin already generates config which heavily uses numeration, you could also take a look there. But please keep the ideas for other plugins, for this only add some docu+unit tests.

File: src/plugins/passwd/CMakeLists.txt (hunk 1 above)
Comment (tom-wa) [inline src/plugins/passwd/CMakeLists.txt]: not really sure how `COMPILE_DEFINITIONS` can be set with if statements, any hints ?
i swear, i'll read up on cmake soon .....

File: src/plugins/passwd/testmod_passwd.c [hunk 4]
```diff
@@ -22,6 +73,14 @@ int main (int argc, char ** argv)
 
 	init (argc, argv);
 
+#ifdef HAS_FGETPWENT
```
Comment (markus2330) [inline src/plugins/passwd/testmod_passwd.c]: is there no workaround if no FGETPWENT is available?

File: src/plugins/passwd/passwd.c [hunk 5]
```diff
@@ -170,6 +170,7 @@ static int writeKS (KeySet * returned, Key * parentKey, SortBy index)
 		return -1;
 	}
 	Key * cur;
+	ksRewind (returned);
```
Comment (markus2330) [inline src/plugins/passwd/passwd.c]: Is this only for unit tests? maybe comment it.

File: src/plugins/passwd/README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+This plugin parses `passwd` files, e.g. `/etc/passwd`. If present, the not-posix complient `fgetpwent` function will be used to read the file supplied by the resolver, if not, `getpwent` will be used. For writing, if present, `putpwent` will be used, if not a simple implementation writing straight to the config file.
+
+## Configuration ##
+
+If the config key `index` is set to `name` passwd entrys will be sorted by name, if not set or set to `uid` passwd entry will be sorted by uid
+
+## Fields ##
+
+`gecos` contains the full name of the account
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: you can itemize the fields (put - in the beginning)

File: src/plugins/passwd/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+- infos/provides = storage
+- infos/recommends =
+- infos/placements = getstorage setstorage
+- infos/status = recommended productive maintained reviewed conformant compatible coverage specific unittest shelltest tested nodep libc configurable final preview memleak experimental difficult unfinished old nodoc concept orphan obsolete discouraged -1000000
+- infos/metadata =
+- infos/description =
+
+## Introduction ##
+
+This plugin parses `passwd` files, e.g. `/etc/passwd`. If present, the not-posix complient `fgetpwent` function will be used to read the file supplied by the resolver, if not, `getpwent` will be used. For writing, if present, `putpwent` will be used, if not a simple implementation writing straight to the config file.
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: From this I would say it should always work (and no ifdef in the unit test should be needed).

Everything but the first sentence is implementation detail, not Introduction.

Comment (tom-wa) [inline src/plugins/passwd/README.md]: the problem is that `getpwent` will only read `/etc/passwd` while with fgetpwent takes a filepointer / can read any file (e.g. `/etc/passwd-`) 
the best solution would probably be parsing the file manually when fgetpwent is not found.

File: src/plugins/passwd/CMakeLists.txt (hunk 1 above)
Comment (markus2330) [inline src/plugins/passwd/CMakeLists.txt]: You do not need an `if` statement, its should be ok if no string (empty string) is after  `COMPILE_DEFINITIONS`.

File: src/error/specification [hunk 8]
```diff
@@ -888,3 +888,10 @@ severity:error
 ingroup:plugin
 module:gitresolver
 macro:GITRESOLVER_CONFLICT
+
+number:149
+description:validaten of passwd entry failed
```
Comment (markus2330) [inline src/error/specification]: typo: validation?

File: src/error/specification [hunk 9]
```diff
@@ -888,3 +888,10 @@ severity:error
...
+severity:error
+ingroup:plugin
+module:passwd
+macro:PASSWD_VALIDATION_ERROR
```
Comment (markus2330) [inline src/error/specification]: is `_ERROR` necessary?

File: src/plugins/passwd/README.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
+- infos = Information about the passwd plugin is in keys below
+- infos/author = Name <name@libelektra.org>
+- infos/licence = BSD
+- infos/needs =
+- infos/provides = storage
+- infos/recommends =
+- infos/placements = getstorage setstorage
+- infos/status = recommended productive maintained reviewed conformant compatible coverage specific unittest shelltest tested nodep libc configurable final preview memleak experimental difficult unfinished old nodoc concept orphan obsolete discouraged -1000000
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: Please fix the status, remove everything that is not appropriate.

File: src/plugins/passwd/README.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+This plugin parses `passwd` files, e.g. `/etc/passwd`. 
+
+## Implementation Details ##
+
+If present, the not-posix complient `fgetpwent` function will be used to read the file supplied by the resolver, if not, `getpwent` will be used. For writing, if present, `putpwent` will be used, if not a simple implementation writing straight to the config file.
+
+## Configuration ##
+
+If the config key `index` is set to `name` passwd entrys will be sorted by name, if not set or set to `uid` passwd entry will be sorted by uid
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: typo: entries

File: src/plugins/passwd/README.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+- infos/metadata =
+- infos/description =
+
+## Introduction ##
+
+This plugin parses `passwd` files, e.g. `/etc/passwd`. 
+
+## Implementation Details ##
+
+If present, the not-posix complient `fgetpwent` function will be used to read the file supplied by the resolver, if not, `getpwent` will be used. For writing, if present, `putpwent` will be used, if not a simple implementation writing straight to the config file.
```
Comment (markus2330) [inline src/plugins/passwd/README.md]: complient -> compliant
`if not` -> otherwise

File: src/plugins/passwd/passwd.c [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+{
+	if (!pwd->pw_name) return -1;
+	if (strlen (pwd->pw_name) == 0) return -1;
+	if (pwd->pw_name[0] == '-') // Posix.1-208 3.431 User Name
+		return -1;
+	const char * invalidCharacters =
+		"/:;<=>?@[\\]^`"; // POSIX.1-2008 3.278 Portable File Character Set - invalid characters > 45 && <= 122
+	for (char * ptr = pwd->pw_name; *ptr != '\0'; ++ptr)
+	{
+		if ((*ptr < 45) || (*ptr > 122) || (strchr (invalidCharacters, *ptr) != NULL)) return -1;
```
Comment (markus2330) [inline src/plugins/passwd/passwd.c]: why 45, 122? prefer to use char-literals

File: src/plugins/passwd/passwd.c [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+#include <sys/stat.h>
+
+#define ID_MAX_CHARACTERS 11
+
+typedef enum {
+	NAME,
+	UID,
+} SortBy;
+
+static int validatepwent (struct passwd * pwd)
```
Comment (markus2330) [inline src/plugins/passwd/passwd.c]: please comment functions

**Comment by markus2330** [comment]
I considered fixing the typos myself, but `infos/status` convinced me not do it it. Please always keep meta data up-to-date, otherwise its worthless.

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
The problem with fast build job was that cmake now does not detect changes in error specification correctly anymore, see #730

Please fix the comments (and ideally also add struct config), then this should be ready to merge.

**Comment by markus2330** [comment]
Thank you, looks nice now! (ignore failing -fast)

Please flag with ready to merge, if finished.

**Comment by markus2330** [comment]
Thank you! Very well done! The plugin is a great addition to Elektra's ecosystem.

**Pull request merged**